src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          if (key in currentState) {
+            const { [key]: _, ...rest } = currentState;
+
+            currentState = rest;
+          }
```

Comment:
The check 'if (key in currentState)' is unnecessary because deleting a non-existent property does not throw an error. Also, you are creating a new copy of the currentState inside the loop for each key, which is inefficient. Instead, you should create a single copy of the currentState and delete the keys from that copy.